clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```
@@ -305,6 +327,52 @@ private boolean flushNetOutBuffer() throws IOException {
         return netOutBuffer.completed();
     }
 
+    private void handleSaslToken(byte[] clientToken) throws SaslException, IOException {
+        if (!enableKafkaSaslAuthenticateHeaders) {
+            byte[] response = saslServer.evaluateResponse(clientToken);
+            if (response != null) {
+                netOutBuffer = new NetworkSend(connectionId, ByteBuffer.wrap(response));
+                flushNetOutBufferAndUpdateInterestOps();
+            }
+        } else {
+            ByteBuffer requestBuffer = ByteBuffer.wrap(clientToken);
+            RequestHeader header = RequestHeader.parse(requestBuffer);
+            ApiKeys apiKey = header.apiKey();
+            short version = header.apiVersion();
+            RequestContext requestContext = new RequestContext(header, connectionId, clientAddress,
+                    KafkaPrincipal.ANONYMOUS, listenerName, securityProtocol);
+            RequestAndSize requestAndSize = requestContext.parseRequest(requestBuffer);
+            if (apiKey != ApiKeys.SASL_AUTHENTICATE) {
+                this.error = Errors.ILLEGAL_SASL_STATE;
+                IllegalSaslStateException e = new IllegalSaslStateException("Unexpected Kafka request of type " + apiKey + " during SASL authentication.");
+                sendKafkaResponse(requestContext, requestAndSize.request.getErrorResponse(e));
+                throw e;
+            }
+            if (!Protocol.apiVersionSupported(apiKey.id, version)) {
+                this.error = Errors.UNSUPPORTED_VERSION;
+                UnsupportedVersionException e = new UnsupportedVersionException("Version " + version + " is not supported for apiKey " + apiKey);
+                sendKafkaResponse(requestContext, requestAndSize.request.getErrorResponse(e));
```

Same question as the above. This doesn't guarantee that all bytes to the client will be flushed before the exception is thrown. Also, if the version is not supported, it's not clear if we could generate a proper response. So, probably we should just throw the exception?

---

Agree, updated.